### PR TITLE
Add types

### PR DESCRIPTION
**Pull request by guillaumegarcia13**
[Typescript newbie here 👶🏻]
Not too sure about this one... 😬
Would happily be reviewed

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,87 +1,91 @@
...
+        "Eugene Sharygin <eush77@gmail.com>",
+        "Titus Wormer <tituswormer@gmail.com> (https://wooorm.com)"
+    ],
+    "files": [
+        "index.js",
+        "types/index.d.ts"
+    ],
+    "types": "types/index.d.ts",
+    "dependencies": {
+        "dtslint": "^4.0.7",
```
Comment (ChristianMurphy) [inline package.json]:
```suggestion
        "dtslint": "^4.0.0",
```

File: types/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+ *
+ * @param tree      Tree to filter
+ * @param options   Whether to drop parent nodes if they had children, but all their children were filtered out
+ * @param test      is-compatible test (such as a type)
+ */
+declare function remove(
+  tree: Node,
+  options: remove.RemoveOptions = {cascade: true},
+  test?: Test<Node>
+): Node;
```
Comment (ChristianMurphy) [inline types/index.d.ts]: I believe this is overloaded there is a second form where test is the second parameter
https://github.com/syntax-tree/unist-util-remove#use

```suggestion
declare function remove(
  tree: Node,
  test?: Test<Node>
): Node;
declare function remove(
  tree: Node,
  options: remove.RemoveOptions = {cascade: true},
  test?: Test<Node>
): Node;
```

File: types/index.d.ts [hunk 3]
```diff
@@ -0,0 +1,26 @@
+// TypeScript Version: 3.5
+
+import { Node } from 'unist';
+import { Test } from 'unist-util-is';
+
+declare namespace remove {
+  interface RemoveOptions {
+    cascade?: boolean;
```
Comment (ChristianMurphy) [inline types/index.d.ts]:
```suggestion
    /**
     * Whether to drop parent nodes if they had children, but all their children were filtered out
test
     */
    cascade?: boolean;
```

File: types/index.d.ts (hunk 2 above)
Comment (ChristianMurphy) [reply types/index.d.ts]: Also the return value will either be the same type that came in, or a null value

```suggestion
declare function remove<T extends Node>(
  tree: T,
  options: remove.RemoveOptions = {cascade: true},
  test?: Test<Node>
): T | null;
```

**Review by ChristianMurphy — COMMENTED** [review]
Thanks @guillaumegarcia13!
It would be good to have a type test for this, similar to https://github.com/stefanprobst/xast-util-sitemap/blob/2a1a85a33cbd5a565f4c7fef7f142160a596bd6d/types/test.ts which was added in https://github.com/syntax-tree/xast-util-sitemap/pull/1

**Comment by guillaumegarcia13** [comment]
Hi @ChristianMurphy!
Thank you very much for the guidance.
Hopefully, it's OK now. 😅

File: types/index.d.ts [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+ * @param tree      Tree to filter
+ * @param options   Whether to drop parent nodes if they had children, but all their children were filtered out
+ * @param test      is-compatible test (such as a type)
+ */
+declare function remove(tree: Node, test?: Test<Node>): Node
+declare function remove(
+  tree: Node,
+  options: remove.RemoveOptions = {cascade: true},
+  test?: Test<Node>
+): Node
```
Comment (ChristianMurphy) [inline types/index.d.ts]: The output will either be the same node type or a null value

```suggestion
declare function remove<T extends Node>(tree: T, test?: Test<Node>): T | null
declare function remove<T extends Node>(
  tree: T,
  options: remove.RemoveOptions = {cascade: true},
  test?: Test<Node>
): T | null
```

File: package.json [hunk 5]
```diff
@@ -1,87 +1,94 @@
...
+        ]
     }
-  },
-  "remarkConfig": {
-    "plugins": [
-      "preset-wooorm"
-    ]
-  }
-}
+}
```
Comment (ChristianMurphy) [inline package.json]: indent and missing newline at end of file seems unexpected.
it doesn't seem to match with prettier?

File: types/tsconfig.json [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+    "include": [
+      "./**/*.ts",
+      "./node_modules/@types"
+    ],
+    "exclude": [
+      "node_modules",
+      "coverage"
+    ]
+  }
+
```
Comment (ChristianMurphy) [inline types/tsconfig.json]: there seems to be an extra level of indent on lines 2-25

File: types/tslint.json [hunk 7]
```diff
@@ -0,0 +1,9 @@
+{
+    "extends": "dtslint/dtslint.json",
+    "rules": {
+        "no-redundant-jsdoc": false,
+        "no-redundant-jsdoc-2": true,
+        "semicolon": false,
+        "whitespace": false
+    }
+}
```
Comment (ChristianMurphy) [inline types/tslint.json]: indent in here seems a bit off, usually it's 2 spaces with a newline at the end of the file

**Review by ChristianMurphy — COMMENTED** [review]
Thanks for following up @guillaumegarcia13! Just a few more tweaks and this should be good to go!

**Comment by ChristianMurphy** [comment]
Would it be possible to have type tests for this PR?
Similar to https://github.com/stefanprobst/xast-util-sitemap/blob/2a1a85a33cbd5a565f4c7fef7f142160a596bd6d/types/test.ts which was added in syntax-tree/xast-util-sitemap#1

File: package.json (hunk 5 above)
Comment (guillaumegarcia13) [reply package.json]: Indeed, .prettierignore is ignoring *.json
I guess it takes my default then (which is 4 spaces)

Comment (wooorm) [reply package.json]: That’s weird! Feel free to remove `*.json` from `.prettierignore` if that fixes things for you

Comment (wooorm) [reply package.json]: > I guess it takes my default then (which is 4 spaces)

We also have an `.editorconfig`, preferring 2 spaces. Most text editors can take those into account. So you might be able to configure your editor to solve it too

Comment (guillaumegarcia13) [reply package.json]: 1. Installed editorconfig globally `npm install -g editorconfig`
2. Added the editorconfig extension to Visual Studio Code
3. Disable extension JS-CSS-HTML Formatter (which was overriding formatting on save)
Works much better now.

File: package.json [hunk 8]
```diff
@@ -30,12 +30,18 @@
...
   ],
   "files": [
-    "index.js"
+    "index.js",
+    "types/index.d.ts"
+  ],
+  "types": "types/index.d.ts",
+  "ignores": [
+    "types"
   ],
```
Comment (ChristianMurphy) [inline package.json]: "files" is an include list, so an exclude list like "ignores" probably isn't needed
```suggestion
```

File: types/index.d.ts [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+}
+
+/**
+ *
+ * Mutate the given tree by removing all nodes that pass test. The tree is walked in preorder (NLR), visiting the node itself, then its head, etc.
+ *
+ * @param tree      Tree to filter
+ * @param test      is-compatible test (such as a type)
+ */
+export function remove(tree: Node, test?: Test<Node>): Node
```
Comment (ChristianMurphy) [inline types/index.d.ts]:
```suggestion
export function remove<T extends Node>(tree: T, test?: Test<Node>): T | null
```

File: types/index.d.ts [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+ *
+ * @param tree      Tree to filter
+ * @param options   Whether to drop parent nodes if they had children, but all their children were filtered out. Default is {cascade: true}
+ * @param test      is-compatible test (such as a type)
+ */
+export function remove(
+  tree: Node,
+  options?: remove.RemoveOptions,
+  test?: Test<Node>
+): Node
```
Comment (ChristianMurphy) [inline types/index.d.ts]:
```suggestion
export function remove<T extends Node>(
  tree: T,
  options?: remove.RemoveOptions,
  test?: Test<Node>
): T | null
```

**Review by ChristianMurphy — APPROVED** [review]

File: types/test.ts [hunk 11]
```diff
@@ -0,0 +1,19 @@
+import {remove} from 'uninst-util-remove';
+import * as u from 'unist-builder';
+
+let tree = u('root', [
```
Comment (ChristianMurphy) [inline types/test.ts]:
```suggestion
const tree = u('root', [
```

**Review by wooorm — APPROVED** [review]

File: types/index.d.ts [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+export function remove<T extends Node>(tree: T, test?: Test<Node>): T | null
+/**
+ *
+ * Mutate the given tree by removing all nodes that pass test. The tree is walked in preorder (NLR), visiting the node itself, then its head, etc.
+ *
+ * @param tree      Tree to filter
+ * @param options   Whether to drop parent nodes if they had children, but all their children were filtered out. Default is {cascade: true}
+ * @param test      is-compatible test (such as a type)
+ */
+export function remove<T extends Node>(
```
Comment (remcohaszing) [inline types/index.d.ts]: This function is exported using `module.exports = remove`.

To reflect this correctly in the types, this should be exported using:

```ts
export = remove
```

File: types/index.d.ts [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+
+// NOTE: namespace is needed to use `export = remove`
+declare namespace remove {}
+
+declare interface RemoveOptions {
+  /**
+   * Whether to drop parent nodes if they had children, but all their children were filtered out test
+   */
+  cascade?: boolean
+}
```
Comment (remcohaszing) [inline types/index.d.ts]: This should be moved into the namespace, so this type can be used as well.

```suggestion
declare namespace remove {
  interface RemoveOptions {
    /**
     * Whether to drop parent nodes if they had children, but all their children were filtered out test
     */
    cascade?: boolean
  }
}
```

Apart from that, looks good to me :+1:

File: types/index.d.ts [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+ *
+ * Mutate the given tree by removing all nodes that pass test. The tree is walked in preorder (NLR), visiting the node itself, then its head, etc.
+ *
+ * @param tree      Tree to filter
+ * @param options   Whether to drop parent nodes if they had children, but all their children were filtered out. Default is {cascade: true}
+ * @param test      is-compatible test (such as a type)
+ */
+declare function remove<T extends Node>(
+  tree: T,
+  options?: RemoveOptions,
```
Comment (remcohaszing) [inline types/index.d.ts]: This is related to the suggestion above.

```suggestion
  options?: remove.RemoveOptions,
```

**Review by remcohaszing — CHANGES_REQUESTED** [review]

**Review by remcohaszing — APPROVED** [review]

**Comment by wooorm** [comment]
Thanks all, released in [2.1.0](https://github.com/syntax-tree/unist-util-remove/releases/tag/2.1.0)!

**Comment by guillaumegarcia13** [comment]
Thank you all for your patience and advice. 🙏🏻
Sorry for all the "noise" generated by my numerous commits. 🙇🏻‍♂️

**Pull request merged**